### PR TITLE
Skip rendering gui nodes with alpha==0

### LINKED ISSUE #6706: Do not render node in GUI if alpha == 0
**Task (REQUIRED):**
Skip node rendering if Alpha == 0

**Expected outcome (REQUIRED):**
If user hide nodes in GUI by setting alpha to 0 it's still rendered on screen and affects fillrate

### PR DESCRIPTION
**Pull request by JCash**
Also skips rendering gui bones.

### CODE REVIEW

File: engine/gui/src/gui.cpp [hunk 1]
```diff
@@ -1107,9 +1107,14 @@ Result DeleteDynamicTexture(HScene scene, const dmhash_t texture_hash)
 
     struct RenderEntrySortPred
     {
-        HScene m_Scene;
-        RenderEntrySortPred(HScene scene) : m_Scene(scene) {}
+        bool operator ()(const RenderEntry& a, const RenderEntry& b) const
```
Comment (JCash) [inline engine/gui/src/gui.cpp:1110]: The scene was never used.

File: engine/gui/src/gui.cpp [hunk 2]
```diff
@@ -1447,14 +1452,31 @@ Result DeleteDynamicTexture(HScene scene, const dmhash_t texture_hash)
...
-            const RenderEntry& entry = c->m_RenderNodes[i];
+            RenderEntry& entry = c->m_RenderNodes[i];
             uint16_t index = entry.m_Node & 0xffff;
             InternalNode* n = &scene->m_Nodes[index];
             float opacity = 1.0f;
+
+            // Ideally, we'd like to be able to do this in an Update step,
+            // but since the gui script is updated _after_ the gui component, we need to accomodate
+            // for any late scripting changes
+            // Note: We need this update step for bones as well, since they update their transform
```
Comment (JCash) [inline engine/gui/src/gui.cpp:1466]: Sadly, this is actually the update for the rendered nodes, and we need this for the gui bones to end up in the correct place (see extension-spine example scene where we attach objects to the gui bones)

File: engine/gui/src/gui.cpp [hunk 3]
```diff
@@ -1447,14 +1452,31 @@ Result DeleteDynamicTexture(HScene scene, const dmhash_t texture_hash)
...
+
+            // Ideally, we'd like to have this update step in the Update function (I'm not even sure why it isn't tbh)
+            // But for now, let's prune the list here
+            if (opacity == 0.0f || n->m_Node.m_IsBone)
+            {
+                entry.m_Node = INVALID_HANDLE;
+                entry.m_RenderKey = 0xFFFFFFFF;
+                ++num_pruned;
+                continue;
+            }
```
Comment (JCash) [inline engine/gui/src/gui.cpp:1478]: Tag the items for removal from this array.

File: engine/gui/src/gui.cpp [hunk 4]
```diff
@@ -1477,17 +1499,16 @@ Result DeleteDynamicTexture(HScene scene, const dmhash_t texture_hash)
             }
         }
 
+        if (num_pruned)
+        {
+            std::sort(c->m_RenderNodes.Begin(), c->m_RenderNodes.End(), RenderEntrySortDisabledPred());
```
Comment (JCash) [inline engine/gui/src/gui.cpp:1504]: Sorting these tiny structs really takes no time at all.
We move all the renderkey=0xFFFFFFFF to the end of the array.

File: engine/gui/src/gui.cpp [hunk 5]
```diff
@@ -1477,17 +1499,16 @@ Result DeleteDynamicTexture(HScene scene, const dmhash_t texture_hash)
...
+            c->m_RenderNodes.SetSize(c->m_RenderNodes.Size() - num_pruned);
```
Comment (JCash) [inline engine/gui/src/gui.cpp:1505]: Here we make the array the correct size, removing the last entries (which we previously sorted there).

File: engine/profiler/src/profile_render.cpp [hunk 6]
```diff
@@ -799,6 +799,8 @@ namespace dmProfileRender
 
     void DeleteProfilerFrame(ProfilerFrame* frame)
     {
+        if (!frame)
```
Comment (JCash) [inline engine/profiler/src/profile_render.cpp:802]: Founnd an issue when shutting down the engine, after the profiler wasn't started properly.

File: engine/gui/src/test/test_gui_script.cpp [hunk 7]
```diff
@@ -49,12 +49,15 @@ static dmLuaDDF::LuaSource* LuaSourceFromStr(const char *str)
 }
 
 void GetTextMetricsCallback(const void* font, const char* text, float width, bool line_break, float leading, float tracking, dmGui::TextMetrics* out_metrics);
+void RenderNodesStoreTransform(dmGui::HScene scene, const dmGui::RenderEntry* nodes, const dmVMath::Matrix4* node_transforms, const float* node_opacities,
```
Comment (JCash) [inline engine/gui/src/test/test_gui_script.cpp:52]: I was a bit annoyed that we had two "RenderScene" functions in the gui.
We should use the one we use in the engine for the tests too.

**Review by britzl — APPROVED** [review]

**Review by AGulev — APPROVED** [review]

**Pull request merged**